### ISSUE TITLE
ReconnectingClientFactory should serialize better

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8057]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#421|
|Type|defect|
|Created|2003-12-05 15:32:08Z|
|Branch|https://github.com/twisted/twisted/tree/reconnecting-client-factory-421-2|

Attachments:

* [p3](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1c7/1c76c42db69f242fdc8d813f5fddf04e958dca4b/e4fbe62d887b8cdee986e6be781203d8d938bbd5) (1935 bytes) - added by warner on 2003-12-05 15:32:07Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__421 421
type__defect defect
reporter__warner warner
priority__highest highest
milestone__ 
branch__branches_reconnecting_client_factory_421_2 branches/reconnecting-client-factory-421-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1070638328000000 1070638328000000
changetime__1469636158984826 1469636158984826
version__ 
owner__ 
cc__itamarst cc__warner
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
When an app that uses RCF is shut down, the connector used by RCF is serialized
along with everything else, and this causes a "Warning: serializing Ephemeral".

I think the correct behavior has two pieces: the connector should never be
serialized (the factory will have to be restarted anyway when the app is
reloaded, so just store a None in the pickle), and the delay/retry state should
be reset to the initial values. This should result in a pickle that, when
reloaded, behaves just like the original RCF did.

I've attached a patch to implement this: please review. The patch also
rearranges stopTrying/stopFactory to do more of the right thing. stopTrying now
does connector.disconnect() (to halt an existing connection *or* a connection
attempt in progress), and clears .continueTrying *before* halting the connection
 (to reduce the chances that it will try to reconnect again right away).
stopFactory is new, and both does .stopTrying and clears the connector (because
once you stopFactory, you have to restart it with a new connector).

Let me know if these look OK and I'll commit them.

thanks,
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
hmm, it appears that my patch breaks test_factories. I don't know why yet.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_3"></a>|@warner set owner to @ralphm|
|-|-|

I've replaced the patch with a smaller one that only addresses the serialization issue (and leaves stopTrying() and stopFactory() as they were). I also added a test case which uses addObserver to watch the log for the warning message that was the original reason for this bug.

These days applications should really be doing enough of their own serialization that they're unlikely to pickle a factory, but I think this patch is still reasonable.

Committed to branches/reconnecting-client-factory-421 for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @warner|
|-|-|

It seems pretty unfortunate to go all the way out to the network just to test the serialization behavior.

How about mocking the reactor and returning an unserializable object as the IConnector?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Ping?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [22516]) Branching to 'reconnecting-client-factory-421-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Wrote something new that doesn't touch the network.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

 * pyflakes:
```
twisted/test/test_factories.py:10: 'util' imported but unused
```
 * trailing whitespaces in protocol.py
 * in `ReconnectingFactoryTestCase`, please remove the tearDown and the self.port variable, and do an addCleanup in testStopTrying instead (this test is pretty bad, that would be great to open a ticket to fix it).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Alright, did all those things.  Also made a couple other minor changes, updated `protocol.py`'s copyright date, removed the "API stability" marker.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [22537]) Merge reconnecting-client-factory-421-2

Author: exarkun
Reviewer: therve
Fixes [#8057](https://github.com/twisted/twisted/issues/8057)

Add tests for the serialization (ie, pickling/unpickling) behavior of ReconnectingClientFactory
and implement a __getstate__ method on that class so that unpickled instances have the same
state as newly created instances (that is, they forget their backoff and retry state).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|GitHub commented|
|-|-|

In changeset 1aac4c147368dcb5cbd1967ccf8bca0c41a45be0
```
#!CommitTicketReference repository="" revision="1aac4c147368dcb5cbd1967ccf8bca0c41a45be0"
Merge pull request #421 from twisted/8691-rodrigc-conch

Author: rodrigc
Reviewer: acabhishek942, einarfd
Fixes: #8691

Python 3 fixes for conch
```
